evm/src/Manager.sol

```diff
@@ -582,7 +718,89 @@ abstract contract Manager is
         emit SiblingUpdated(siblingChainId, oldSiblingContract, siblingContract);
     }
 
+    function endpointAttestedToMessage(bytes32 digest, uint8 index) public view returns (bool) {
+        return _getMessageAttestationsStorage()[digest].attestedEndpoints & uint64(1 << index) == 1;
+    }
+
+    function attestationReceived(
+        uint16 sourceChainId,
+        bytes32 sourceManagerAddress,
+        EndpointStructs.ManagerMessage memory payload
+    ) external onlyEndpoint {
+        _verifySibling(sourceChainId, sourceManagerAddress);
+
+        bytes32 managerMessageHash = EndpointStructs.managerMessageDigest(sourceChainId, payload);
+
+        // set the attested flag for this endpoint.
+        // NOTE: Attestation is idempotent (bitwise or 1), but we revert
+        // anyway to ensure that the client does not continue to initiate calls
+        // to receive the same message through the same endpoint.
+        if (
```

Comment:
`executeMsg` is now a public function, this will serve as a recovery path for when an endpoint is disabled and an attestation has already been delivered by a different endpoint.